src/ralph/virtual/models.py

```diff
@@ -197,6 +200,16 @@ class Meta:
     def __str__(self):
         return self.hostname
 
+    @cached_property
+    def rack_id(self):
+        return self.rack.id if self.rack else None
+
+    @cached_property
+    def rack(self):
+        if isinstance(self.hypervisor, DataCenterAsset):
```

Comment:
hypervisor has to be either DataCenterAsset or None, so simple check `if self.hypervisor` will be enough here